streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2262,31 +2275,39 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
    */
   final def repeatWith[R1 <: R, B, C](
     schedule: Schedule[R1, Any, B]
-  )(f: O => C, g: B => C): ZStream[R1, E, C] =
-    ZStream[R1, E, C] {
+  )(f: O => C, g: B => C): ZStream[R1 with Clock, E, C] =
+    ZStream[R1 with Clock, E, C] {
+      import Schedule.Interval
+      import Schedule.Decision._
+
+      import java.time.Instant
+
       for {
-        scheduleInit  <- schedule.initial.toManaged_
-        schedStateRef <- Ref.make(scheduleInit).toManaged_
+        schedStateRef <- Ref.make(Option.empty[Instant] -> schedule.step).toManaged_
         switchPull    <- ZManaged.switchable[R1, Nothing, ZIO[R1, Option[E], Chunk[C]]]
         currPull      <- switchPull(self.map(f).process).flatMap(as => Ref.make(as)).toManaged_
         doneRef       <- Ref.make(false).toManaged_
         pull = {
-          def go: ZIO[R1, Option[E], Chunk[C]] =
+          def go: ZIO[R1 with Clock, Option[E], Chunk[C]] =
             doneRef.get.flatMap { done =>
               if (done) Pull.end
               else
                 currPull.get.flatten.foldM(
                   {
                     case e @ Some(_) => ZIO.fail(e)
                     case None =>
-                      schedStateRef.get
-                        .flatMap(schedule.update((), _))
-                        .foldM(
-                          _ => doneRef.set(true) *> Pull.end,
-                          state =>
-                            switchPull((self.map(f) ++ ZStream.succeed(g(schedule.extract((), state)))).process)
-                              .tap(currPull.set(_)) *> schedStateRef.set(state) *> go
-                        )
+                      schedStateRef.get.flatMap {
+                        case (start, step) =>
+                          clock.instant.flatMap(now =>
+                            start.fold[URIO[Clock, Any]](ZIO.unit)(start => ZIO.sleep(Interval(now, start).size)) *>
```

Comment:
Here we're sleeping and then stepping. Should it be the other way around? E.g. only sleep when `Continue` is returned?